agora/Agora/Proposal/Time.hs

```diff
@@ -178,6 +179,30 @@ deriving via
 instance AdditiveSemigroup (Term s PPOSIXTime) where
   (punsafeCoerce @_ @_ @PInteger -> x) + (punsafeCoerce @_ @_ @PInteger -> y) = punsafeCoerce $ x + y
 
+{- | Get the starting time of a proposal, from the 'Plutus.V1.Ledger.Api.txInfoValidPeriod' field.
+     For every proposal, this is only meant to run once upon creation.
+-}
+createProposalStartingTime :: forall (s :: S). Term s (PPOSIXTime :--> PPOSIXTimeRange :--> PProposalStartingTime)
+createProposalStartingTime = phoistAcyclic $
+  plam $ \maxDuration iv -> unTermCont $ do
+    currentTimeF <-
+      tcont $
+        pletFields @'["lowerBound", "upperBound"] $
+          currentProposalTime # iv
+
+    -- Use the middle of the current time range as the starting time.
+    let duration = currentTimeF.upperBound - currentTimeF.lowerBound
+
+        startingTime =
+          pdiv
+            # (currentTimeF.lowerBound + currentTimeF.upperBound)
+            # 2
+
+    tcassert "createProposalStartingTime: given time range should be tight enough" $
+      duration #<= maxDuration
+
+    pure $ pcon $ PProposalStartingTime startingTime
```

Comment:
Great stuff. This does make me think: `ProposalTime` ought to be a Scott-encoded type, since it will never live in a datum.

Reply:
Beautiful! Thanks :)